MediaService: Did not call startForeground().

Fenix: https://github.com/mozilla-mobile/fenix/issues/4853

```
RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{df2820b u0 org.mozilla.fenix/mozilla.components.feature.media.service.MediaService}
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1872)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:214)
    at android.app.ActivityThread.main(ActivityThread.java:7050)
    at java.lang.reflect.Method.invoke(Method.java)
...
(2 additional frame(s) were not displayed)

Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{df2820b u0 org.mozilla.fenix/mozilla.components.feature.media.service.MediaService}
```

Filtered log:

```
MediaStateMachine  I  Transitioning from MediaState.None to Playing(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/vid
                            eo/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(PLAY)])
           MediaService  D  Service created
                         D  Command received
             AudioFocus  D  processAudioFocusResult(1)
      MediaStateMachine  I  Transitioning from Playing(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croat
                            ie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(PLAY)]) to Playing(session=Session(420fff9f-b206-419e-b
                            ef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), med
                            ia=[Media(PLAY), Media(PLAY)])
           MediaService  D  Command received
             AudioFocus  D  processAudioFocusResult(1)
      MediaStateMachine  I  Transitioning from Playing(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croat
                            ie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(WAITING), Media(PAUSE)]) to Playing(session=Session(420
                            fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-
                            8136603.php), media=[Media(WAITING)])
           MediaService  D  Command received
             AudioFocus  D  processAudioFocusResult(1)
      MediaStateMachine  I  Transitioning from Playing(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croat
                            ie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(PAUSE)]) to Paused(session=Session(420fff9f-b206-419e-b
                            ef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), med
                            ia=[Media(PAUSE), Media(PAUSE)])
           MediaService  D  Command received
      MediaStateMachine  I  Transitioning from Paused(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croati
                            e-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(ENDED), Media(PAUSE)]) to Paused(session=Session(420fff9
                            f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-8136
                            603.php), media=[Media(PAUSE)])
           MediaService  D  Command received
      MediaStateMachine  I  Transitioning from Paused(session=Session(420fff9f-b206-419e-bef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croati
                            e-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), media=[Media(PAUSE)]) to Playing(session=Session(420fff9f-b206-419e-b
                            ef9-b0cfb5d7aa8c, http://www.leparisien.fr/video/arnaque-en-croatie-la-villa-etait-un-terrain-vague-20-08-2019-8136603.php), med
                            ia=[Media(PLAY)])
         AndroidRuntime  D  Shutting down VM
                         E  FATAL EXCEPTION: main
                         E  Process: org.mozilla.samples.browser, PID: 10119
                         E  android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{
                            f84ab82 u0 org.mozilla.samples.browser/mozilla.components.feature.media.service.MediaService}
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
                         E      at android.os.Handler.dispatchMessage(Handler.java:107)
                         E      at android.os.Looper.loop(Looper.java:214)
                         E      at android.app.ActivityThread.main(ActivityThread.java:7356)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

Now it happened very quickly:

```
MediaStateMachine  I  Transitioning from MediaState.None to Playing(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/vid
                      eo/rentree-ministerielle-connaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PLAYING,playbackState=PLAY)])
     MediaService  D  Service created
                   D  Command received
       AudioFocus  D  processAudioFocusResult(1)
     MediaService  D  startForeground()

                      Process org.mozilla.samples.browser:media created for service {org.mozilla.samples.browser/org.mozilla.gecko.media.MediaManager
                      }
                      PID: 16408   UID:    GIDs:

MediaStateMachine  I  Transitioning from Playing(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-minister
                      ielle-connaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PAUSED,playbackState=PAUSE)]) to Paused(session=
                      Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-connaissez-vous-vos-ministres
                      -21-08-2019-8136792.php), media=[Media(state=PAUSED,playbackState=PAUSE)])
     MediaService  D  Command received
                   D  stopForeground()
MediaStateMachine  I  Transitioning from Paused(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministeri
                      elle-connaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=ENDED,playbackState=ENDED)]) to MediaState.None
     MediaService  D  Command received
   AndroidRuntime  D  Shutting down VM
                   E  FATAL EXCEPTION: main
                   E  Process: org.mozilla.samples.browser, PID: 15915
                   E  android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{
                      9a01536 u0 org.mozilla.samples.browser/mozilla.components.feature.media.service.MediaService}
                   E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
                   E      at android.os.Handler.dispatchMessage(Handler.java:107)
                   E      at android.os.Looper.loop(Looper.java:214)
                   E      at android.app.ActivityThread.main(ActivityThread.java:7356)
                   E      at java.lang.reflect.Method.invoke(Native Method)
                   E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
                   E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

```
--------------
PLAY:
--------------

      MediaStateMachine  I  Transitioning from MediaState.None to Playing(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rent
                            ree-ministerielle-connaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PLAYING,playbackState=PLAY)])
           MediaService  D  Service created
                         D  Command received
             AudioFocus  D  processAudioFocusResult(1)
           MediaService  D  startForeground()

                            Process org.mozilla.samples.browser:media created for service {org.mozilla.samples.browser/org.mozilla.gecko.media.MediaManager}
                            PID: 23962   UID:    GIDs:

      MediaStateMachine  I  Transitioning from Playing(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-c
                            onnaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PAUSED,playbackState=PAUSE)]) to Paused(session=Session(8ca60f
                            a8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-connaissez-vous-vos-ministres-21-08-2019-8136792.p
                            hp), media=[Media(state=PAUSED,playbackState=PAUSE)])
           MediaService  D  Command received
                         D  stopForeground()
      MediaStateMachine  I  Transitioning from Paused(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-co
                            nnaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PLAYING,playbackState=PLAY)]) to Playing(session=Session(8ca60f
                            a8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-connaissez-vous-vos-ministres-21-08-2019-8136792.p
                            hp), media=[Media(state=PLAYING,playbackState=PLAY)])

--------------
REFRESH!
--------------

                         I  Transitioning from Playing(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-c
                            onnaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PAUSED,playbackState=PAUSE)]) to Paused(session=Session(8ca60f
                            a8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-connaissez-vous-vos-ministres-21-08-2019-8136792.p
                            hp), media=[Media(state=PAUSED,playbackState=PAUSE)])
           MediaService  D  Command received
                         D  stopForeground()
      MediaStateMachine  I  Transitioning from Paused(session=Session(8ca60fa8-2cba-41ca-aeb7-ad22c335d9ec, http://www.leparisien.fr/video/rentree-ministerielle-co
                            nnaissez-vous-vos-ministres-21-08-2019-8136792.php), media=[Media(state=PAUSED,playbackState=PAUSE)]) to MediaState.None
           MediaService  D  Command received
         AndroidRuntime  D  Shutting down VM
                         E  FATAL EXCEPTION: main
                         E  Process: org.mozilla.samples.browser, PID: 23641
                         E  android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{c2c9f96
                             u0 org.mozilla.samples.browser/mozilla.components.feature.media.service.MediaService}
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
                         E      at android.os.Handler.dispatchMessage(Handler.java:107)
                         E      at android.os.Looper.loop(Looper.java:214)
                         E      at android.app.ActivityThread.main(ActivityThread.java:7356)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

--- a/components/concept/engine/src/main/java/mozilla/components/concept/engine/media/Media.kt
+++ b/components/concept/engine/src/main/java/mozilla/components/concept/engine/media/Media.kt
@@ -190,7 +190,7 @@ abstract class Media(
         }
     }
 
-    override fun toString(): String = "Media($playbackState)"
+    override fun toString(): String = "Media(state=$state,playbackState=$playbackState)"
 }
 
 /**

--- a/components/feature/media/src/main/java/mozilla/components/feature/media/notification/MediaNotification.kt
+++ b/components/feature/media/src/main/java/mozilla/components/feature/media/notification/MediaNotification.kt
@@ -39,12 +39,19 @@ internal class MediaNotification(
             .setContentTitle(data.title)
             .setContentText(data.description)
             .setLargeIcon(data.largeIcon)
-            .addAction(data.action)
             .setPriority(NotificationCompat.PRIORITY_LOW)
             .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
-            .setStyle(androidx.media.app.NotificationCompat.MediaStyle()
-                .setMediaSession(mediaSession.sessionToken)
-                .setShowActionsInCompactView(0))
+
+        val style = androidx.media.app.NotificationCompat.MediaStyle()
+            .setMediaSession(mediaSession.sessionToken)
+
+        if (data.action != null) {
+            builder.addAction(data.action)
+
+            style.setShowActionsInCompactView(0)
+        }
+
+        builder.setStyle(style)
 
         if (!state.isForExternalApp()) {
             // We only set a content intent if this media notification is not for an "external app"
@@ -89,7 +96,9 @@ private fun MediaState.toNotificationData(context: Context): NotificationData {
                     0)
             ).build()
         )
-        else -> throw IllegalArgumentException("Cannot create notification for state: $this")
+        // Dummy notification that is only used to satisfy the requirement to ALWAYS call
+        // startForeground with a notification.
+        else -> NotificationData()
     }
 }
 
@@ -106,11 +115,11 @@ private val Session.nonPrivateIcon: Bitmap?
     get() = if (private) null else icon
 
 private data class NotificationData(
-    val title: String,
-    val description: String,
-    @DrawableRes val icon: Int,
+    val title: String = "",
+    val description: String = "",
+    @DrawableRes val icon: Int = R.drawable.mozac_feature_media_playing,
     val largeIcon: Bitmap? = null,
-    val action: NotificationCompat.Action
+    val action: NotificationCompat.Action? = null
 )
 
 private fun MediaState.isForExternalApp(): Boolean {

--- a/components/feature/media/src/main/java/mozilla/components/feature/media/service/MediaService.kt
+++ b/components/feature/media/src/main/java/mozilla/components/feature/media/service/MediaService.kt
@@ -53,7 +53,7 @@ internal class MediaService : Service() {
     }
 
     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
-        logger.debug("Command received")
+        logger.debug("Command received: ${intent?.action}")
 
         when (intent?.action) {
             ACTION_UPDATE_STATE -> processCurrentState()
@@ -67,7 +67,9 @@ internal class MediaService : Service() {
 
     private fun processCurrentState() {
         val state = MediaStateMachine.state
+
         if (state == MediaState.None) {
+            updateNotification(state)
             shutdown()
             return
         }
@@ -93,13 +95,21 @@ internal class MediaService : Service() {
 
         val notification = notification.create(state, mediaSession)
 
-        if (state is MediaState.Playing) {
-            startForeground(notificationId, notification)
-        } else {
+        // Android wants us to always, always, ALWAYS call startForeground() if
+        // startForegroundService() was invoked. Even if we already did that for this service or
+        // if we are stopping foreground or stopping the whole service. No matter what. Always
+        // call startForeground().
+        startForeground(notificationId, notification)
+
+        if (state !is MediaState.Playing) {
             stopForeground(false)
 
             NotificationManagerCompat.from(this)
                 .notify(notificationId, notification)
+        }
+
+        if (state is MediaState.None) {
+            stopForeground(true)
         }
     }
 

--- a/components/feature/media/src/test/java/mozilla/components/feature/media/notification/MediaNotificationTest.kt
+++ b/components/feature/media/src/test/java/mozilla/components/feature/media/notification/MediaNotificationTest.kt
@@ -17,7 +17,6 @@ import mozilla.components.support.test.robolectric.testContext
 import org.junit.Assert.assertEquals
 import org.junit.Test
 import org.junit.runner.RunWith
-import java.lang.IllegalArgumentException
 
 @RunWith(AndroidJUnit4::class)
 class MediaNotificationTest {
@@ -59,14 +58,20 @@ class MediaNotificationTest {
         assertEquals(R.drawable.mozac_feature_media_paused, notification.iconResource)
     }
 
-    @Test(expected = IllegalArgumentException::class)
     fun `media notification for none state`() {
-        // This notification will actually never get displayed
+        // This notification is only used to call into startForeground() to fulfil the API contract.
+        // It gets immediately replaced or removed.
 
         val state = MediaState.None
 
         MediaNotification(testContext)
             .create(state, mock())
+
+        val notification = MediaNotification(testContext)
+            .create(state, mock())
+
+        assertEquals("", notification.text)
+        assertEquals("", notification.title)
     }
 
     @Test

